### PR TITLE
Unify linkage for intel_adsp: No more "bootloader"

### PR DESCRIPTION
**Pull request by andyross**
Big series.   Not quite ready for merge yet, because it sits on top of the more focused changes in https://github.com/zephyrproject-rtos/zephyr/pull/41045

The core feature here is that the build and link of the "bootloader" stage is now unified with the main Zephyr link, meaning we can see IMR memory space as linked area where we can put functions and data (currently just initialization, but anything with an __imr attribute specified will be put there) and make calls across the boundary.

Along the way, the CPU initialization step is unified for all cores, and we remove a bunch of extra steps due to using the Zephyr arch "reset vector" code (we don't handle reset -- that happens in a ROM) along with SOF-derived initialization paths (mostly descended from the same Cadence code I think) and some Intel hardware dependencies that were always in the bootloader C code.

In the process, the main image .text shrinks by a total of ~1500 bytes and the code by a total of about 1000 lines (mostly due to duplicated/redundant startup code above).  I'm really happy with how this turned out.

As mentioned I'll put a DNM on this until the dependencies merge, and I expect a few rounds of reorganization along the way.  It's ready for review though, so please have at it.

### CODE REVIEW

File: soc/xtensa/intel_adsp/common/include/cpu_init.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	 * inaccessible (15) for safety.  Note that there is a HAL
+	 * routine that does this (by emulating the older "cacheattr"
+	 * hardware register), but it generates significantly larger
+	 * code.
+	 */
+	if (IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V25)) {
+		/* Already set up by the ROM on older hardware. */
+		const uint8_t attribs[] = { 2, 15, 15, 15, 2, 4, 15, 15 };
+
+		for (int region = 0; region < 8; region++) {
```
Comment (peter-mitsis) [inline soc/xtensa/intel_adsp/common/include/cpu_init.h]: I am surprised to see a hardcoded value of 8 here. Granted, it makes complete sense--it matches the number of elements in <attribs>. I just expected to see something like ...

    for (int region = 0; region < NUM_ELEMENTS(attribs); region++) {

... however, I am unable to find such a macro (perhaps I just missed it by another name).

Anyway, this most definitely is not a show stopper; it is just an observation that jumped out at me.

Comment (andyross) [reply soc/xtensa/intel_adsp/common/include/cpu_init.h]: Actually this goes away in this PR.  What you're looking at here is just code motion from already-existing code, the newer API is a little cleaner and puts the region attributes into a single line of macro arguments.

But in this case the 8 is firmly architectural.  Xtensa Region Protection Option partitions memory into 8 segments based on the top three bits of the address, and that's all it does.  2^3 == 8.

File: soc/xtensa/intel_adsp/common/include/cavs-mem.h [hunk 2]
```diff
@@ -23,17 +23,6 @@
...
-/* These are fake section addresses used only in the linker scripts */
-#define IDT_BASE		(L2_SRAM_BASE + L2_SRAM_SIZE)
-#define IDT_SIZE		0x2000
-#define UUID_ENTRY_ELF_BASE	0x1FFFA000
-#define UUID_ENTRY_ELF_SIZE	0x6000
-#define LOG_ENTRY_ELF_BASE	0x20000000
-#define LOG_ENTRY_ELF_SIZE	0x2000000
-#define EXT_MANIFEST_ELF_BASE	(LOG_ENTRY_ELF_BASE + LOG_ENTRY_ELF_SIZE)
-#define EXT_MANIFEST_ELF_SIZE	0x2000000
+#define IMR_BOOT_LDR_TEXT_ENTRY_BASE (CONFIG_IMR_MANIFEST_ADDR + 0x6000)
```
Comment (teburd) [inline soc/xtensa/intel_adsp/common/include/cavs-mem.h:26]: What is the 0x6000 here do we know? Can we define it to make it easier to understand?

File: soc/xtensa/intel_adsp/common/include/cpu_init.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * Note that cAVS 1.5 has the "translation" option that we don't use,
+ * but still need to put an identity mapping in the high bits.  Also
+ * per spec changing the current code region requires that WITLB be
+ * followed by an ISYNC and that both instructions live in the same
+ * cache line (two 3-byte instructions fit in an 8-byte aligned
+ * region, so that's guaranteed not to cross a caceh line boundary).
+ */
+#define SET_ONE_TLB(region, att) do {					\
+	uint32_t addr = region * 0x20000000U, attr = att;		\
```
Comment (teburd) [inline soc/xtensa/intel_adsp/common/include/cpu_init.h:30]: can we do a define for the region size here?

File: soc/xtensa/intel_adsp/common/include/cpu_init.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	 * writeback caching (4) in the sixth mapping (the second of
+	 * two RAM mappings) and to mark all unused regions
+	 * inaccessible (15) for safety.  Note that there is a HAL
+	 * routine that does this (by emulating the older "cacheattr"
+	 * hardware register), but it generates significantly larger
+	 * code.
+	 */
+	SET_RPO_TLB(2, 15, 15, 15, 2, 4, 15, 15);
+
+	/* Initialize ATOMCTL: Hardware defaults for S32C1I use
```
Comment (teburd) [inline soc/xtensa/intel_adsp/common/include/cpu_init.h:111]: I appreciate all the great commentary in here, it makes it understandable even to me, and I'm certainly not as familiar with xtensa or the adsp yet. Thanks!

File: soc/xtensa/intel_adsp/common/include/soc.h [hunk 5]
```diff
@@ -72,9 +66,62 @@
...
+ * in the L1 cache will not be inspected nor modified by the access.
+ *
+ * @see z_soc_cached_ptr()
+ *
+ * @param p A pointer to a valid C object
+ * @return A pointer to the same object bypassing the L1 dcache
+ */
+static inline void *z_soc_uncached_ptr(void *p)
+{
+	return ((void *)(((size_t)p) & ~0x20000000));
```
Comment (teburd) [inline soc/xtensa/intel_adsp/common/include/soc.h:103]: I maybe misinterpretting here, but the address has a bit that signifies cacheable or uncacheable as I read this, perhaps defining the bit with BIT() and using that define here would be a nice way to make this a little more readable.

File: soc/xtensa/intel_adsp/common/include/soc.h [hunk 6]
```diff
@@ -72,9 +66,62 @@
...
+ *
+ * @see z_soc_uncached_ptr()
+ *
+ * @param p A pointer to a valid C object
+ * @return A pointer to the same object via the L1 dcache
+
+ */
+static inline void *z_soc_cached_ptr(void *p)
+{
+	return ((void *)(((size_t)p) | 0x20000000));
```
Comment (teburd) [inline soc/xtensa/intel_adsp/common/include/soc.h:124]: same comment as above

**Review by teburd — APPROVED** [review]
Some minor nits that might help readability. This looks like a really nice changeset to move the intel_adsp soc to be closer to Zephyr

File: soc/xtensa/intel_adsp/common/include/soc.h (hunk 5 above)
Comment (andyross) [reply soc/xtensa/intel_adsp/common/include/soc.h:103]: (So... this is another comment that actually belongs in the earlier PR, heh.  But no reason not to discuss it here.)

It's actually an offset, not a bit.  The bit math is an optimization.

Xtensa's "region protection option" partitions memory into 8 regions, each of which can be provided different protection settings, including cache control.  The hardware designers mapped the RAM blocks twice, once in 0x80000000 and again at 0xa000000.  And we exploit that so we can put some symbols or regions into one "space" or the other and control their cacheability via what pointer we use to access them (with great care, obviously!).

As it happens those are adjacent so we can effect the offset by toggling bit 0x20000000.  But that's an artifact of the hardware mapping decision.

Other hardware (ahem, watch this space) might make different decisions and (sigh) use a non-power-of-two offset.  And it needs a different abstraction still than you see here.

File: soc/xtensa/intel_adsp/common/include/cavs-mem.h (hunk 2 above)
Comment (andyross) [reply soc/xtensa/intel_adsp/common/include/cavs-mem.h:26]: Rimage places the start of the "TEXT" section it parsed out of the bootloader module 24k after the location of the manifest in the output file, which is itself a rimage-generated data structure.  It's just the way it works, basically.  I can move the magic number around or give it different names but it doesn't stop being magic.  And it lives at exactly one place in the code now.

Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs-mem.h:26]: Fair enough, the rest of this is so well commented this little magic num seems out of place given the lack of commentary, the blurb you've put here would be helpful in code to better understand it the next time (inevitably) someone has to read it

File: soc/xtensa/intel_adsp/common/include/soc.h (hunk 5 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/soc.h:103]: That completely makes sense now, I read the comment here but didn't quite grok it, but the comment does describe just this already I believe. Thanks!

File: soc/xtensa/intel_adsp/common/include/cavs-mem.h (hunk 2 above)
Comment (andyross) [reply soc/xtensa/intel_adsp/common/include/cavs-mem.h:26]: Agreed, I'll see what I can do.  Really this is sort of a wart in rimage.  We should just say "put our boot blob at this address" and it should figure out the padding and give a single yea/nay answer as to whether that fits with the hardware requirements (actually it's CSME software that defines the image format, so ultimately the offsets get designed there, way outside of the world of open source visibility).

File: soc/xtensa/intel_adsp/common/boot.c [hunk 7]
```diff
@@ -297,6 +292,7 @@ static __imr void lp_sram_init(void)
 	}
 
 	CAVS_SHIM.ldoctl = SHIM_LDOCTL_LPSRAM_LDO_BYPASS;
+	bbzero((void *)L2_SRAM_BASE, L2_SRAM_SIZE);
```
Comment (dcpleung) [inline soc/xtensa/intel_adsp/common/boot.c:295]: Shouldn't these be `LP_SRAM_BASE` and `LP_SRAM_SIZE`?

Comment (andyross) [reply soc/xtensa/intel_adsp/common/boot.c:295]: Whoops, yes.

Comment (andyross) [reply soc/xtensa/intel_adsp/common/boot.c]: Fixed.  In fact that exposed the fact we had bad sizes (presumptively bad anyway; either that or there's missing code to enable sram bank somewhere) for the LP-SRAM block.  Useful catch.

**Review by dcpleung — APPROVED** [review]

**Comment by andyross** [comment]
Clean up some checkpatchisms.  Note that the remaining errors are from the inline assembly, and fundamentally false positives.  I've used this pattern everywhere and like it a lot, but checkpatch assumes "strings are for printing" and doesn't like whitespace before a newline.  Note that if I replace the whitespace with two separate string literals it issues a different warning (it doesn't like adjacent literals).  Another option would be to use a semicolon to terminate the expressions instead of newlines, but that breaks the DWARF line numbering and results in unreadable output in zephyr.lst.  Trust me, this is the best way to do this. :)

**Comment by kv2019i** [comment]
@andyross I was about to put this PR for a test run with SOF CI, but it fails already at build phase:
https://sof-ci.01.org/sofpr/PR4914/build11365/build/tgl_zph_error.txt
/srv/home/jenkins/xcc/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-ld: error: no memory region specified for loadable section `.module.boot'

Same error for other cAVS versions, here's cavs15:
https://sof-ci.01.org/sofpr/PR4914/build11365/build/apl_zph_error.txt

**Comment by andyross** [comment]
OK, thanks.  xcc was working for me earlier, but the last few revs I haven't been trying it.  Clearly I broke something; will fix.

**Comment by andyross** [comment]
Yeah, it was that last rush of linker cleanup that did it.  You'd have thought binutils was mostly unchanged since the 90's, but it turns out that a handful of nice features have landed in the past few years.  Undid the accidental dependencies on recent objcopy, basically.  Left it as a separate patch for easier review.

**Review by kv2019i — APPROVED** [review]
Thanks now passes SOF test suite (did a local test on 2-core TGL-H in addition).

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
There's one legitimate error in the compliance check failure:
```
Commit e35c923311:
1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Unify Xtensa CPU reset between the secondary cores and the main core."
```
Otherwise this PR looks ready to go!

**Review by MaureenHelm — APPROVED** [review]

**Review by lyakh — COMMENTED** [review]
Very impressive, huge thanks! Almost a 1000 line code size reduction! At least one question - why does the whole LP SRAM have to be zeroed?

**Pull request merged**